coderd/provisionerdserver/provisionerdserver.go

```diff
@@ -194,6 +222,46 @@ func (s *server) timeNow() time.Time {
 	return dbtime.Now()
 }
 
+// heartbeat runs heartbeatOnce at the interval specified by HeartbeatInterval
+// until the lifecycle context is canceled.
+func (s *server) heartbeat() {
+	tick := time.NewTicker(time.Nanosecond)
+	defer tick.Stop()
+	for {
+		select {
+		case <-s.lifecycleCtx.Done():
+			return
+		case <-tick.C:
+			hbCtx, hbCancel := context.WithTimeout(s.lifecycleCtx, s.HeartbeatInterval)
+			if err := s.heartbeatOnce(hbCtx); err != nil {
+				if !xerrors.Is(err, context.DeadlineExceeded) && !xerrors.Is(err, context.Canceled) {
+					s.Logger.Error(hbCtx, "heartbeat failed", slog.Error(err))
+				}
+			}
+			hbCancel()
+			tick.Reset(s.HeartbeatInterval)
+		}
+	}
+}
+
+// heartbeatOnce updates the last seen at timestamp in the database.
+// If HeartbeatFn is set, it will be called instead.
+func (s *server) heartbeatOnce(ctx context.Context) error {
```

Comment:
I'm not sure if I caught the reason why this function has the "Once" suffix. Usually, it signals `sync.Once`, but not in this case.